reccmp/isledecomp/compare/queries.py

```diff
@@ -105,3 +101,42 @@ def get_floats_without_data(
             yield (orig_addr, size == 8)
         elif image_id == ImageId.RECOMP and isinstance(recomp_addr, int):
             yield (recomp_addr, size == 8)
+
+
+def get_referencing_entity_matches(db: EntityDb) -> Iterator[tuple[int, int]]:
+    """Return new matches for child entities that refer to the same parent entity.
+    These can be import thunks, incremental build thunks, or vtordisps.
+    To match, the child entities must have the same displacement values (or none).
+    If we cannot match uniquely, match by child address order in each address space.
+    """
+    for orig_addr, recomp_addr in db.sql.execute(
+        """
+        WITH linked_refs AS (
+            SELECT r.img, r.addr, m.match_id, disp0, disp1,
+            Row_number() OVER (partition BY r.img, m.match_id, disp0, disp1 order by r.addr) nth
+            FROM refs r
+            -- Convert the referenced address to a unique ID to allow for matching.
+            INNER JOIN matches m
+                ON (r.img = 0 AND r.ref = m.addr_x)
+                OR (r.img = 1 AND r.ref = m.addr_y)
+            -- Exclude thunk entities that have been matched.
+            INNER JOIN (
+                SELECT img, addr FROM refs
+                EXCEPT
+                SELECT img, addr FROM matched_ids
+            ) x
+            ON r.img = x.img AND r.addr = x.addr
+        )
+        SELECT x.addr, y.addr FROM
+        (SELECT * FROM linked_refs WHERE img = 0) x
+        INNER JOIN
+        (SELECT * FROM linked_refs WHERE img = 1) y
+        ON  x.disp0 = y.disp0
+        AND x.disp1 = y.disp1
+        AND x.nth = y.nth
+        AND x.match_id = y.match_id
```

Comment:
1. Could you put all SQL keywords in uppercase (e.g. `PARTITION BY`, `ORDER BY`)?
2. I spent a few minutes and I think I understood the gist of it, but this statement is still quite complex. I'll partially trust in your experience, the outcome, and the tests. Please let me know in case you want a more detailed review.
3. I haven't thought too much about whether the test coverage by the existing and new tests is sufficient. Do you think this function would benefit from dedicated unit tests?